core/src/main/java/com/microsoft/applicationinsights/internal/heartbeat/HeartBeatProvider.java

```diff
@@ -0,0 +1,304 @@
+ *  {@link com.microsoft.applicationinsights.internal.heartbeat.HeartBeatProviderInterface} and
+ *  {@link com.microsoft.applicationinsights.internal.shutdown.Stoppable}
+ * </p>
+ *
+ * @author Dhaval Doshi
+ * @since 03-30-2018
+ */
+public class HeartBeatProvider implements HeartBeatProviderInterface, Stoppable {
+
+  /**
+   * The name of the heartbeat metric.
+   */
+  private final String HEARTBEAT_SYNTHETIC_METRIC_NAME = "HeartbeatState";
+
+  /**
+   * The list of disabled properties
+   */
+  private List<String> disableDefaultProperties = new ArrayList<>();
+
+  /**
+   * List of disabled heartbeat providers
+   */
+  private List<String> disabledHeartBeatPropertiesProviders = new ArrayList<>();
+
+  /**
+   * The counter for heartbeat sent to portal
+   */
+  private AtomicLong heartbeatsSent;
+
+  /**
+   * Map to hold heartbeat properties
+   */
+  private ConcurrentMap<String, HeartBeatPropertyPayload> heartbeatProperties;
+
+  /**
+   * Interval at which heartbeat would be sent
+   */
+  private long interval;
+
+  /**
+   * Telemetry client instance used to send heartbeat.
+   */
+  private TelemetryClient telemetryClient;
+
+  /**
+   * ThreadPool used for adding properties to concurrent dictionary
+   */
+  private ExecutorService propertyUpdateService;
+
+  /**
+   * Threadpool used to send data heartbeat telemetry
+   */
+  private ScheduledExecutorService heartBeatSenderService;
+
+  /**
+   * Heartbeat enabled state
+   */
+  private volatile boolean isEnabled;
+
+  private final Object lock = new Object();
+
+  public HeartBeatProvider() {
+    this.interval = HeartBeatProviderInterface.DEFAULT_HEARTBEAT_INTERVAL;
+    this.heartbeatProperties = new ConcurrentHashMap<>();
+    this.isEnabled = true;
+    this.heartbeatsSent = new AtomicLong(0);
+    this.propertyUpdateService = Executors.newCachedThreadPool(ThreadPoolUtils.createDaemonThreadFactory(HeartBeatProvider.class));
+    this.heartBeatSenderService = Executors.newScheduledThreadPool(1, ThreadPoolUtils.createDaemonThreadFactory(HeartBeatProvider.class));
+  }
+
+  @Override
+  public String getInstrumentationKey() {
+    return this.telemetryClient.getContext().getInstrumentationKey();
+  }
+
+  @Override
+  public void setInstrumentationKey(String key) {
+    if (this.telemetryClient != null && key != null) {
+      this.telemetryClient.getContext().setInstrumentationKey(key);
+    }
+  }
+
+  @Override
+  public void initialize(TelemetryConfiguration configuration) {
+    if (isEnabled) {
+      if (this.telemetryClient == null) {
+        this.telemetryClient = new TelemetryClient(configuration);
+      }
+
+      //Submit task to set properties to dictionary using separate thread. we do not wait for the
+      //results to come out as some I/O bound properties may take time.
+      propertyUpdateService.submit(HeartbeatDefaultPayload.populateDefaultPayload(getExcludedHeartBeatProperties(),
```

Comment:
Which properties are I/O bound? I wasn't able to find any. We may be able to remove the thread pool.

Reply:
There are no properties which are IO bound today. But later on there will be when we will query ARM to get VM ID. I would rather keep it.

Reply:
Makes sense.